ql/src/test/org/apache/hadoop/hive/ql/optimizer/calcite/TestCBOReCompilation.java

```diff
@@ -0,0 +1,115 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.optimizer.calcite;
+
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.conf.HiveConf.ConfVars;
+import org.apache.hadoop.hive.ql.DriverFactory;
+import org.apache.hadoop.hive.ql.IDriver;
+import org.apache.hadoop.hive.ql.processors.CommandProcessorException;
+import org.apache.hadoop.hive.ql.session.SessionState;
+import org.apache.hive.testutils.HiveTestEnvSetup;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+public class TestCBOReCompilation {
+
+  @ClassRule
+  public static HiveTestEnvSetup env_setup = new HiveTestEnvSetup();
+
+  @BeforeClass
+  public static void beforeClass() throws Exception {
+    try (IDriver driver = createDriver()) {
+      dropTables(driver);
+      String[] cmds = {
+          // @formatter:off
+          "create table aa1 ( stf_id string)",
+          "create table bb1 ( stf_id string)",
+          "create table cc1 ( stf_id string)",
+          "create table ff1 ( x string)"
+          // @formatter:on
+      };
+      for (String cmd : cmds) {
+        driver.run(cmd);
+      }
+    }
+  }
+
+  @AfterClass
+  public static void afterClass() throws Exception {
+    try (IDriver driver = createDriver()) {
+      dropTables(driver);
+    }
+  }
+
+  public static void dropTables(IDriver driver) throws Exception {
+    String[] tables = new String[] {"aa1", "bb1", "cc1", "ff1" };
+    for (String t : tables) {
+      driver.run("drop table if exists " + t);
+    }
+  }
+
+  @Test
+  public void testReExecutedOnError() throws Exception {
+    try (IDriver driver = createDriver("ALWAYS")) {
+      String query = "explain from ff1 as a join cc1 as b " +
+          "insert overwrite table aa1 select   stf_id GROUP BY b.stf_id " +
+          "insert overwrite table bb1 select b.stf_id GROUP BY b.stf_id";
+      driver.run(query);
+    }
+  }
+
+  @Test
+  public void testFailOnError() throws Exception {
+    try (IDriver driver = createDriver("TEST")) {
+      String query = "explain from ff1 as a join cc1 as b " +
+          "insert overwrite table aa1 select   stf_id GROUP BY b.stf_id " +
+          "insert overwrite table bb1 select b.stf_id GROUP BY b.stf_id";
+      Assert.assertThrows("Plan not optimized by CBO", CommandProcessorException.class, () -> driver.run(query));
```

Comment:
these are bad tests - because it kinda sets in stone that we should not fix this issue

if we let this in - then you will force who will fix this issue to also invent a new query for this which:
* fails in cbo
* doesn not fail without cbo

I don't think that's fair - so a different test is needed

but I understand that this is something which should be handle....right now I don't have any good ideas what we could do instead of this - I'll keep thinking about it...

Reply:
Found other tests created by @zabetak, so no need to keep these